### PR TITLE
feat(TPG >=4.23)!: Add docker_registry variables and use them in google_cloudfunctions_function resource

### PR DESCRIPTION
**Pull request by tardigrde**
Variables added: 
- docker_registry
- docker_repository 
- kms_key_name

This enables storing Cloud Build container in Artifact Registry

See https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloudfunctions_function#docker_registry

### CODE REVIEW

**Comment by tardigrde** [comment]
Hi! Could you please look at this merge request? I know I created and closed some PRs with the same scope, those were problematic ecause of failing CLAs. Now the commit is with an email address that signed a CLA.

Happy to improve the PR if needed. Let me know if I have to do anything else.

The important thing for me was that I can use Artifact Registry for storing Cloud Build image. I think no default functionality was changed.

Can you please run `/gcbrun`?  I think I do not have the rights to do it.


Lets't try

/gcbrun

**Comment by apeabody** [comment]
Thanks for the contribution @tardigrde - Relevant output from the LINT tests:

```
Checking for documentation generation
diff -r '--exclude=.terraform' '--exclude=.kitchen' '--exclude=autogen' '--exclude=*.tfvars' '--exclude=*metadata.yaml' /workspace/README.md /tmp/tmp.7eUSZEwJ41/generate_docs/workspace/README.md
64c64
< | docker\_repository | User managed repository created in Artifact Registry optionally with a customer managed encryption key. If specified, deployments will use Artifact Registry. | `string` | n/a | yes |
---
> | docker\_repository | User managed repository created in Artifact Registry optionally with a customer managed encryption key. If specified, deployments will use Artifact Registry. | `string` | `""` | no |
71c71
< | kms\_key\_name | Resource name of a KMS crypto key (managed by the user) used to encrypt/decrypt function resources. | `string` | n/a | yes |
---
> | kms\_key\_name | Resource name of a KMS crypto key (managed by the user) used to encrypt/decrypt function resources. | `string` | `""` | no |
Error: Documentation generation has not been run, please run the
'make docker_generate_docs' command and commit the above changes.
Checking for trailing whitespace
Checking for missing newline at end of file
Error: No newline at end of file ./variables.tf
```

**Comment by tardigrde** [comment]
Hey! I fixed the linting problems, can you please do gcbrun again? 

See commit: 4ee3f29417c0b93140bdb6d334a1857ee557605f

**Comment by tardigrde** [comment]
Hi @apeabody ! The checks have been passed, now I tried to updae the branch, but again a gcbrun is needed. Can you please review my changes and merge if no problems found. Thanks!

File: variables.tf [hunk 1]
```diff
@@ -184,3 +184,22 @@ variable "build_environment_variables" {
...
+
+variable "docker_registry" {
+  type        = string
+  default     = "CONTAINER_REGISTRY"
+  description = "Docker Registry to use for storing the function's Docker images. Allowed values are CONTAINER_REGISTRY (default) and ARTIFACT_REGISTRY."
+}
+
+variable "docker_repository" {
+  type        = string
+  default     = ""
```
Comment (apeabody) [inline variables.tf:196]: Would this be better as `null`?

```suggestion
  default     = null
```

**Review by apeabody — CHANGES_REQUESTED** [review]
Thanks for the contribution @tardigrde!

I've included a few comments below, and additionaly it doesn't appear all these fields are present in TPG 4.11, so https://github.com/terraform-google-modules/terraform-google-event-function/blob/master/versions.tf#L23 will need to be bumped to an appropriate minimum version.

Thanks!

File: variables.tf [hunk 2]
```diff
@@ -184,3 +184,22 @@ variable "build_environment_variables" {
   default     = {}
   description = "A set of key/value environment variable pairs available during build time."
 }
+
+variable "docker_registry" {
+  type        = string
+  default     = "CONTAINER_REGISTRY"
```
[the remark this answers is not in the document]
Comment (tardigrde) [reply variables.tf:190]: You are right. This should be null.

**Comment by tardigrde** [comment]
Thanks for the heads up! I'm still learning..

Bump google provider min version to v4.23
v4.22 does not have the parameters yet, but v4.23 does have them, see versioned doc: https://registry.terraform.io/providers/hashicorp/google/4.23.0/docs/resources/cloudfunctions_function

**Comment by apeabody** [comment]
> Ha! I should probably pay closer attention to open PRs. #165 is pretty much a copy of this, as I need this feature as well.
> 
> @apeabody @tardigrde Any way we can move it forward?

Triggered the CI

**Comment by apeabody** [comment]
Hi @tardigrde - When you have a chance please run `make docker_generate_docs`, thanks!

**Comment by tardigrde** [comment]
Updated! Anything else?

**Review by apeabody — APPROVED** [review]
Thanks for the contribution @tardigrde!

**Comment by tardigrde** [comment]
@apeabody would you be so kind to make a release? I'd like to get this [PR](https://github.com/terraform-google-modules/terraform-google-scheduled-function/pull/120) merged too, and for that I need a release in this repo (it's a dependency to scheduled-function module)

**Comment by apeabody** [comment]
> @apeabody would you be so kind to make a release? I'd like to get this [PR](https://github.com/terraform-google-modules/terraform-google-scheduled-function/pull/120) merged too, and for that I need a release in this repo (it's a dependency to scheduled-function module)

Hi @tardigrde - Sure, please track the release at #167

**Pull request merged**